src/ServerScriptService/Services/BulletHandlerService/init.lua

```diff
@@ -1,14 +1,228 @@
+--!strict
+
 local ReplicatedStorage = game:GetService("ReplicatedStorage")
 local ServerScriptService = game:GetService("ServerScriptService")
 local CollectionService = game:GetService("CollectionService")
 local Players = game:GetService("Players")
 
 local Knit = require(ReplicatedStorage.Packages.Knit)
 
+local GunDataTypes = require(ReplicatedStorage.Common.Types.GunDataTypes)
+local BulletHistory = require(script.BulletHistory)
+
+type validationDataType = {
+    Tool: Tool,
+    BulletData: GunDataTypes.BulletData,
+    GunConfig: GunDataTypes.GunConfig,
+    BinderInstance: any
+}
+
 local BulletHandlerService = Knit.CreateService {
     Name = "BulletHandlerService";
     Client = {};
 }
+BulletHandlerService.DEBUG_MODE = true
+
+local damageMultiplierList = {
+    ["Head"] = 2,
+    ["UpperTorso"] = 1.1,
+    ["LowerTorso"] = 0.9,
+
+    ["RightUpperLeg"] = 0.8,
+    ["RightLowerLeg"] = 0.8,
+    ["RightFoot"] = 0.8,
+    ["LeftUpperLeg"] = 0.8,
+    ["LeftLowerLeg"] = 0.8,
+    ["LeftFoot"] = 0.8,
+
+    ["RightUpperArm"] = 0.7,
+    ["RightLowerArm"] = 0.7,
+    ["RightHand"] = 0.7,
+    ["LeftUpperArm"] = 0.7,
+    ["LeftLowerArm"] = 0.7,
+    ["LeftHand"] = 0.7
+}
+
+function BulletHandlerService:_log(str: string, ...): nil
+    if not self.DEBUG_MODE then
+        return
+    end
+    print("[BulletHandlerService] " .. str:format(...))
+end
+
+function BulletHandlerService:GetRaycastBlacklist(): Array<Instance>
+	local blacklist = CollectionService:GetTagged("Character")
```

Comment:
It checks if there's walls/structures between the bullet start and endPoint, it should ignore characters because they've already moved a lot by the time the shot is taken on the original client